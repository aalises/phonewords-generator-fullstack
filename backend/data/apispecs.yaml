basePath: "/api/v1"
schemes:
  - http
info:
  description: "Simple API to convert phone numbers to phone words. The base URL for retrieving the resources is /api/v1"
  title: "PhoneWords API"
  version: 1.0.0
externalDocs:
  description: "Project repository"
  url: "https://github.com/aalises/phonewords-generator-fullstack"  
produces: 
    - "application/json"    
paths:
  "/phonewords/<number>":
    get:
      summary: Computes all the phonewords associated with a number
      tags: 
        - resources
      produces:
        - application/json  
      description: Computes all the phonewords associated with a number
      responses: 
          "200":
            description: "Phone word response with a list of all possible phonewords, success flag and error message if applicable"
            schema:
              $ref: '#/definitions/Phonewords'
          default:
            description: "Unexpected Error"

definitions:
  Phonewords:
    required:
      - success
      - phonewords
      - error
    properties:
      success:
        type: boolean
        description: Success flag of the call
      phonewords:
        type: array
        items:
          type: string
        description: List with all the possible phonewords
      error:
        type: string
        description: Error message       
          
swagger: '3.0'