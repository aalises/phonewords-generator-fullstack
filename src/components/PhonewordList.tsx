import React, { Fragment } from 'react'
import styled from "styled-components"

import Heading from "@kiwicom/orbit-components/lib/Heading";
import Separator from "@kiwicom/orbit-components/lib/Separator";
import Alert from "@kiwicom/orbit-components/lib/Alert";
import { ClipLoader } from "react-spinners";

interface PhonewordListProps {
    number: string,
    data: PhoneWordResponse,
    isSubmitting: boolean
}

const Number = styled.span`
  border: ${props => props.number.length ? '1px solid #D3D3D3' : 'none'}
  border-radius: 3px
  padding: 0.3rem
  color: #666666
  cursor: default
`

const WordTag = styled.div`
  font-size: 14px
  margin: 1rem 0.1rem
  cursor: default
  display: inline-block
  border: 1px solid
  padding: 6px 8px
  border-radius: 3px
  color: white
  background: #666666
  font-weight: 500
`

export default ({number, data, isSubmitting} : PhonewordListProps) => {
  const {success, words, error} = data

  //Different alerts for errors on the request or no words
  const Alerts = () => {
    if(number){
      if(!words.length) return <Alert title={null} icon>There are no words matching your request</Alert>
      if(!success) return <Alert type="critical" title={null} icon>{error ? error : 'Unknown Error'}</Alert>     
    }
    return null
  }

  return (
    <div>
        <Heading element="h2" type="title2" spaceAfter="medium">
          {number ? 
            <Fragment>List of Phone Words for the number: <Number number={number}>{number}</Number></Fragment> : 
            <Fragment>Submit a Phone Number to display the list of its possible phone words</Fragment>
          }
        </Heading>
        <Separator />
        {isSubmitting ? 
          <ClipLoader size={20} color={'grey'} /> : 
          <Fragment> 
            {words.map(word => <WordTag key={word}>{word}</WordTag>)}
            <Alerts/>
          </Fragment>
        }
    </div>
  )
}
