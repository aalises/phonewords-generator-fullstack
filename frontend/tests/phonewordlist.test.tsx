import * as React from 'react';
import { shallow } from 'enzyme';
import PhonewordList from '../src/components/PhonewordList';
import data from "./data";

describe("Phoneword List Test Suite", () => {

  const createListRenderer = (props) => {
    return shallow(<PhonewordList {...props} />);
  }
    
  const checkSnapshot = (data) => {
    const list = createListRenderer(data);
    expect(list).toMatchSnapshot();
  }

  it('Renders the Initial state correctly', () => {
    checkSnapshot(data.initial);
  });
  
  it('Renders the Alert message when no words correctly', () => {
    checkSnapshot(data.alert);
  });
  
  it('Displays a spinner when submitting', () => {
    checkSnapshot(data.submitting);
  });
  
  it('Displays a list of words from a specific number as expected', () => {
    checkSnapshot(data.number);
  });
});