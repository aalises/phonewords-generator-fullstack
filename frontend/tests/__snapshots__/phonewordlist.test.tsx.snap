// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Phoneword List Test Suite Displays a list of words from a specific number as expected 1`] = `
<div>
  <Heading
    dataTest="heading-phoneword-list"
    element="h2"
    spaceAfter="medium"
    type="title2"
  >
    List of Phone Words for the number: 
    <styled.span
      data-test="phoneword-number"
      number="66743"
    >
      66743
    </styled.span>
  </Heading>
  <Separator />
  <styled.div
    data-test="word-tag-test"
    key="test"
  >
    test
  </styled.div>
  <styled.div
    data-test="word-tag-phonewords"
    key="phonewords"
  >
    phonewords
  </styled.div>
  <Alerts />
</div>
`;

exports[`Phoneword List Test Suite Displays a spinner when submitting 1`] = `
<div>
  <Heading
    dataTest="heading-phoneword-list"
    element="h2"
    spaceAfter="medium"
    type="title2"
  >
    Submit a Phone Number to display the list of its possible phone words
  </Heading>
  <Separator />
  <onlyUpdateForKeys(Loader)
    color="grey"
    css=""
    loading={true}
    size={20}
    sizeUnit="px"
  />
</div>
`;

exports[`Phoneword List Test Suite Renders the Alert message when no words correctly 1`] = `
<div>
  <Heading
    dataTest="heading-phoneword-list"
    element="h2"
    spaceAfter="medium"
    type="title2"
  >
    Submit a Phone Number to display the list of its possible phone words
  </Heading>
  <Separator />
  <Alerts />
</div>
`;

exports[`Phoneword List Test Suite Renders the Initial state correctly 1`] = `
<div>
  <Heading
    dataTest="heading-phoneword-list"
    element="h2"
    spaceAfter="medium"
    type="title2"
  >
    Submit a Phone Number to display the list of its possible phone words
  </Heading>
  <Separator />
  <Alerts />
</div>
`;
